gitAndTools.overcommit: init at 0.46.0

Gerschtli opened the pull request:
###### Motivation for this change

Init overcommit at 0.46.0.

Added `vendor` in `.gitignore` because this directory gets generated while building `gemset.nix`.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Gerschtli commented:
To use overcommit, it will copy some files (namely `template-dir/hooks/*`) into the `.git/hooks` directory of a git repo where you want to install it. The problem is that the gem overcommit is required in these scripts (via `require 'overcommit`) but can't be found. I have no idea how to fix this. Updating the `LOAD_PATH` in gemset-config doesn't work..

Does anyone have an idea?

Gerschtli committed:
gitAndTools.overcommit: init at 0.46.0

Gerschtli commented:
The problem is that ruby needs to be installed along with overcommit to get this working. How can I tell nix to do so?

alyssais commented on pkgs/applications/version-management/git-and-tools/overcommit/.bundle/config:
```diff
@@ -0,0 +1,3 @@
+---
+BUNDLE_PATH: "vendor/bundle"
+BUNDLE_CACHE_ALL: "true"
```
Why do we need this file?

Gerschtli replied:
We don't, I have removed this file

alyssais commented on pkgs/applications/version-management/git-and-tools/overcommit/default.nix:
```diff
@@ -0,0 +1,15 @@
+{ lib, bundlerEnv, ruby }:
+
+bundlerEnv {
```
This should probably be `bundlerApp`.

Gerschtli replied:
With a `bundlerApp` the provided hooks won't work because ruby needs to find the overcommit library.

alyssais committed:
overcommit: let installed hooks find the library

Patches overcommit to use the bundlerApp's GEM_HOME and GEM_PATH. This
means nothing has to be installed globally to use overcommit. A side
effect of this is that, if overcommit is installed globally, hooks will
continue to use the version of overcommit that generated them, and will
have to be regenerated to use a different version of overcommit. This
seems reasonable to me.

Gerschtli commented:
It does not work for me..

1. I checked out the branch.
2. Executed `nix run -f . gitAndTools.overcommit mdl ruby`.
3. Created minimal config:
```yml
verify_signatures: false

PreCommit:
  Mdl:
    enabled: true
```
4. Executed `overcommit --install`
5. Staged a change in a Markdown file.
6. Then tried to commit, resulting in this:
```
This repository contains hooks installed by Overcommit, but the `overcommit` gem is not installed.
Install it with `gem install overcommit`.
```

Furthermore, if I understand your changes correctly, the hooks aren't wrapped:
```
$> head -n20 .git/hooks/pre-commit                                                                                                                                                                                               
#!/nix/store/k6bmp8wix8f350nwkrmcybx3b7yfbcam-ruby-2.5.3/bin/ruby                                                                                                                                                                

# Entrypoint for Overcommit hook integration. Installing Overcommit will result
# in all of your git hooks being copied from this file, allowing the framework
# to manage your hooks for you.

# Prevent a Ruby stack trace from appearing when we interrupt the hook.
# Note that this will be overridden when Overcommit is loaded, since the
# InterruptHandler will redefine the trap at that time.
Signal.trap('INT') do
  puts 'Hook run interrupted'
  exit 130
end

# Allow hooks to be disabled via environment variable so git commands can be run
# in scripts without Overcommit running hooks
if ENV['OVERCOMMIT_DISABLE'].to_i != 0 || ENV['OVERCOMMIT_DISABLED'].to_i != 0
  exit
end
```

Gerschtli commented:
@alyssais Are you still interested in giving your input? Otherwise I will close this PR because I am unable to fix this.

alyssais commented:
> @alyssais Are you still interested in giving your input? Otherwise I will close this PR because I am unable to fix this.

I am, yes, but I have a lot going on at the moment, so will likely be at
least a couple of weeks before I can get to it.

Gerschtli commented:
As I'm not able to fix this, I'm closing this PR. If anyone wants to reopen and fix it, feel free to do so.
